File: src/game/BaseFighter.ts
import type { FighterConfig, FighterId, FighterPalette, AttackStyle, NativeFacing } from "./types";

export abstract class BaseFighter {
  abstract readonly id: FighterId;
  abstract readonly name: string;
  abstract readonly greeting: string;
  abstract readonly signatureAttack: string;
  abstract readonly attackStyle: AttackStyle;
  abstract readonly power: number;
  abstract readonly speed: number;
  abstract readonly defense: number;
  abstract readonly criticalChance: number;
  abstract readonly baseDamage: number;
  abstract readonly scale: number;
  abstract readonly nativeFacing: NativeFacing;
  abstract readonly palette: FighterPalette;

  toConfig(): FighterConfig {
    return {
      id: this.id,
      name: this.name,
      greeting: this.greeting,
      signatureAttack: this.signatureAttack,
      attackStyle: this.attackStyle,
      power: this.power,
      speed: this.speed,
      defense: this.defense,
      criticalChance: this.criticalChance,
      baseDamage: this.baseDamage,
      scale: this.scale,
      nativeFacing: this.nativeFacing,
      palette: this.palette
    };
  }
}

File: csharp-reference/Models/BaseFighter.cs
namespace DragonBallArena.Models;

/// <summary>
/// VSPets-style character definition. The VS Code runtime mirrors these values
/// in TypeScript because VS Code extensions execute JavaScript/TypeScript.
/// </summary>
public abstract class BaseFighter(string? name = null)
{
    public string Name { get; } = name ?? string.Empty;
    public abstract string Id { get; }
    public abstract string DisplayName { get; }
    public abstract string Greeting { get; }
    public abstract string SignatureAttack { get; }
    public abstract string AttackStyle { get; }
    public abstract double Scale { get; }
    public abstract FighterStats Stats { get; }
    public abstract FighterPalette Palette { get; }
}
